### PR TITLE
Add Google login flow with sessions

### PR DESCRIPTION
**Pull request by kq-li**
## Overview

I implemented the Google login flow with cornell.edu emails only and added a simple session system with a session token valid for one day and an update token that can be used to refresh the session.

## Changes Made

I set up the OAuth flow from scratch because I felt that the original implementation with Flask-OAuth was difficult to follow and obfuscated, whereas this is pretty straightforward and extensible.

## Test Coverage

I tested this locally, since it mainly involves interacting with the Google identity platform.

### CODE REVIEW

**Comment by kq-li** [comment]
lol ok let's merge #8 first and i can make this a controller while i implement the DB integration

File: app.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from flask import Flask, redirect, request
+import os
```
Comment (young-k) [inline app.py]: shouldn't python imports go before other library imports?

Comment (kq-li) [reply app.py]: ye

File: src/app/coursegrab/controllers/oauth2_callback_controller.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from os import environ
+import requests
```
Comment (manishsaha) [inline src/app/coursegrab/controllers/oauth2_callback_controller.py]: imports go first right?

File: src/app/coursegrab/controllers/oauth2_callback_controller.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+from . import *
+from app.coursegrab.dao.users_dao import create_user
```
Comment (manishsaha) [inline src/app/coursegrab/controllers/oauth2_callback_controller.py]: this is already imported in `from . import *`

File: src/app/coursegrab/controllers/oauth2_callback_controller.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        token_json = token_res.json()
+        if not token_res.ok:
+            raise Exception("Error fetching token: " + token_json["error_description"])
+
+        id_res = requests.get("https://oauth2.googleapis.com/tokeninfo", params={"id_token": token_json["id_token"]})
+        id_json = id_res.json()
+        if not id_res.ok:
+            raise Exception("Error fetching token info: " + id_json["error_description"])
+
+        created, user = create_user(id_json["email"], id_json["given_name"], id_json["family_name"])
```
Comment (manishsaha) [inline src/app/coursegrab/controllers/oauth2_callback_controller.py]: I think you have to do `users_dao.create_user(...)` here with above change

File: src/app/coursegrab/models/user.py [hunk 5]
```diff
@@ -11,15 +14,38 @@
...
         self.last_name = kwargs.get("last_name")
+        self.courses = db.relationship("Course", secondary=users_to_courses, backref="users")
+        self.refresh_session()
+
+    def generate_token(self):
+        return hashlib.sha1(os.urandom(64)).hexdigest()
+
+    def refresh_session(self):
+        self.session_token = self.generate_token()
+        self.session_expiration = datetime.datetime.now() - datetime.timedelta(days=1)
```
Comment (manishsaha) [inline src/app/coursegrab/models/user.py]: i am confused here, why is the session_expiration subtracting from the current datetime?

Comment (kq-li) [reply src/app/coursegrab/models/user.py]: aiya this was debug

File: src/app/coursegrab/controllers/oauth2_callback_controller.py (hunk 2 above)
Comment (kq-li) [reply src/app/coursegrab/controllers/oauth2_callback_controller.py]: system imports before third party i think is the rule

File: src/app/coursegrab/utils/__init__.py [hunk 6]
```diff
@@ -1 +1,3 @@
 from app.coursegrab.utils.appdev_controller import AppDevController  # noqa: F401
+from app.coursegrab.utils.appdev_redirect_controller import AppDevRedirectController  # noqa: F401
```
Comment (yuna878) [inline src/app/coursegrab/utils/__init__.py]: You could probably use `# flake8: noqa` at the top of file to ignore the entire file

File: src/app/coursegrab/__init__.py [hunk 7]
```diff
@@ -1,12 +1,22 @@
...
+from app.coursegrab.controllers.login_controller import *
+from app.coursegrab.controllers.oauth2_callback_controller import *
 from app.coursegrab.controllers.track_course_controller import *
+from app.coursegrab.controllers.update_session_controller import *
 
 # CourseGrab Blueprint
 coursegrab = Blueprint("coursegrab", __name__, url_prefix="/api")
 
-controllers = [HelloWorldController(), CreateUserController(), TrackCourseController()]
+controllers = [
```
Comment (yuna878) [inline src/app/coursegrab/__init__.py]: What does controller mean? Is it a standard to naming specific type of classes?

Comment (manishsaha) [reply src/app/coursegrab/__init__.py]: standard for naming files which house the endpoint logic

**Pull request merged**